File: 695-max-area-of-island/max-area-of-island.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int maxArea = 0;
        for(int i = 0; i < grid.length; i++) {
            for(int j = 0;j < grid[0].length; j++) {
                if(grid[i][j] != 0 && grid[i][j] != -1) {
                    maxArea = Math.max(maxArea,calculateArea(grid,0,i,j)/4);
                }
            }
        }
        return maxArea;
    }

    public int calculateArea(int[][] map, int area,int r, int c) {
        int count = 0;
        if(r < 0 || r >= map.length || c < 0 || c >= map[0].length) {
            return area;
        }
        if(map[r][c] == 0 || map[r][c] == -1) {
            return area;
        }
        int[][] dir = new int[][]{{0,1},{1,0},{0,-1},{-1,0}};
        for(int[] directions: dir) {
            map[r][c] = -1;
            int val = calculateArea(map,area + 1,r + directions[0],c + directions[1]);
            area = Math.max(area,val);
        }
        return area;
    }
}

File: 695-max-area-of-island/max-area-of-island.py
class Solution(object):
    dirs = [[0,1],[0,-1],[1,0],[-1,0]]

    def maxAreaOfIsland(self, grid):
        visited = []
        for i in range(len(grid)):
            row = [False] * len(grid[0])
            visited.append(row)
        
        maxArea = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1 and not visited[i][j]:
                    maxArea = max(maxArea,self.dfs(grid,i,j,visited))
            
        return maxArea

    def dfs(self,grid,row,column,visited):
        if grid[row][column] == 0:
            return 0
        
        visited[row][column] = True

        total = 0

        for direction in self.dirs:
            nr = row + direction[0]
            nc = column + direction[1]
            if nr >= 0 and nr < len(grid) and nc >= 0 and nc < len(grid[0]):
                if not visited[nr][nc]:
                    total += self.dfs(grid,nr,nc,visited)
        
        return total + 1        